int32 targetPos
int32 speed